### PR TITLE
PR: Add pysyntect recipe

### PR DESCRIPTION
**Pull request by andfoy**
This PR adds [pysyntect](https://github.com/spyder-ide/pysyntect) to conda-forge. Currently, this package depends on Rust nightly, which is not available on conda-forge, however, we don't need it for execution, as we produce a static library based on C.

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml"
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example)
- [x] Source is from official source
- [ ] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged)
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details)
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there

### CODE REVIEW

**Comment by andfoy** [comment]
Pinging @ccordoba12 to ask for his permission to participate on this package

**Comment by ccordoba12** [comment]
> https://community.plotly.com/t/plotly-express-in-spyder-offline-instead-of-iplot/21743

I agree to be part of the maintainers team for this feedstock.

**Comment by andfoy** [comment]
@ccordoba12, this one is ready at last!

**Comment by ccordoba12** [comment]
Great!! Thanks a lot for hard work on this one @andfoy!

@conda-forge/staged-recipes, could you help us with this one? This library provides Python bindings for [sytenct](https://github.com/trishume/syntect) Rust library for syntax highlighting. That's why the installation process is kind of convoluted.

**Comment by mariusvniekerk** [comment]
Wow that's quite involved

**Comment by isuruf** [comment]
How about adding a nightly branch to `rust-feedstock` and use it from there?

**Comment by andfoy** [comment]
> How about adding a nightly branch to rust-feedstock and use it from there?

Could you please help us opening a ``dev`` (CFEP-05) branch there, it seems that it does not exist right now

**Comment by andfoy** [comment]
> Since this is building a static library using dependencies with licenses like Apache-2.0, you need to distribute the licenses and copyright notices for all of the dependencies.

Syntect is MIT license, should I include it also?

**Comment by isuruf** [comment]
> Syntect is MIT license, should I include it also?

That doesn't mean that the static binary is MIT license. For eg: `pyo3-derive-backend` is Apache-2.0. Therefore you need to include the license of `pyo3-derive-backend` in this package. Same issue with the 91 other dependencies used. Maybe rust has a way of giving you the license of all the packages.

**Comment by ccordoba12** [comment]
@isuruf, could you create a `dev` branch in the rust feedstock so that @andfoy can create a PR for nightly against it?

**Comment by isuruf** [comment]
@andfoy, can you create a PR against master of `rust-feedstock` and ask a maintainer to merge it into a new branch?

**Comment by isuruf** [comment]
https://github.com/maghoff/cargo-license-hound might help you with the license issue.

**Comment by andfoy** [comment]
> @andfoy, can you create a PR against master of `rust-feedstock` and ask a maintainer to merge it into a new branch?

Sure! I'll do it

**Comment by andfoy** [comment]
@isuruf Do you know why Rust nightly is not being picked up on Python 3.6 builds?

**Comment by isuruf** [comment]
For windows, can you send a PR to remove this condition checking at https://github.com/conda-forge/staged-recipes/blob/master/.ci_support/build_all.py#L34 ? It's not needed.

**Comment by andfoy** [comment]
> For windows, can you send a PR to remove this condition checking at https://github.com/conda-forge/staged-recipes/blob/master/.ci_support/build_all.py#L34 ? It's not needed.

Sure!

**Comment by andfoy** [comment]
@isuruf I think that the Windows fix worked. But I don't know why the recipe is not pulling rust-dev on Python 3.6

**Comment by isuruf** [comment]
The line at https://github.com/conda-forge/staged-recipes/blob/master/.ci_support/build_all.py#L56 needs to be guarded with a `if not built` condition as well.

**Comment by andfoy** [comment]
The last PR fixes this recipe, thanks @isuruf ! I'll add the missing licenses and this should be ready to go

**Comment by andfoy** [comment]
@isuruf Are all the licenses correct? Or should they belong on other folder?

**Comment by isuruf** [comment]
For Apache licenses you need the copyright notice too. (not just the license text). License text does not have to be duplicated.

**Comment by andfoy** [comment]
If a library provides both Apache and MIT licenses, is it good to stick to only one of those?

**Comment by isuruf** [comment]
I don't mind either. If you are sticking with one, use MIT.

**Comment by andfoy** [comment]
> I don't mind either. If you are sticking with one, use MIT.

Then I'll stick to MIT, it seems that many libraries didn't post their COPYRIGHT notices

**Comment by ccordoba12** [comment]
@isuruf, I think this is ready. Or is there something else @andfoy needs to address here?

**Comment by isuruf** [comment]
No, it's not. Please see my comment above.

**Comment by andfoy** [comment]
@isuruf I think we are ready!

**Comment by isuruf** [comment]
Why is the license of `clang-sys` deleted?

**Comment by andfoy** [comment]
> Why is the license of clang-sys deleted?

They don't have the COPYRIGHT posted and no MIT license

**Comment by isuruf** [comment]
They do have copyright notices at https://github.com/KyleMayes/clang-sys/blob/master/src/lib.rs#L1
You'll need to distribute them, otherwise you can't distribute `pysyntect` in binary form.

**Comment by andfoy** [comment]
Done!

**Comment by isuruf** [comment]
Is that the only `Apache-2.0` dependency?

**Comment by andfoy** [comment]
> Is that the only Apache-2.0 only dependency?

Yes, at least by the inspection I did

**Comment by isuruf** [comment]
Okay, I'll have a look too. Can you add code to check that each dependency used in the build has a file of the form `<pkg>-LICENSE*`? Otherwise we won't notice with new versions.

File: recipes/pysyntect/build.sh [hunk 1]
```diff
@@ -0,0 +1,24 @@
+set -ex
+
+# Set conda CC as custom CC in Rust
+export CARGO_TARGET_X86_64_UNKNOWN_LINUX_GNU_LINKER=$CC
+
+# Print Rust version
+rustc --version
+
+# Install cargo-license
+cargo install cargo-license
```
Comment (isuruf) [inline recipes/pysyntect/build.sh:10]: Can you install this into `$BUILD_PREFIX` instead of `$HOME/.cargo/bin`?

Comment (andfoy) [reply recipes/pysyntect/build.sh:10]: Let me check

**Comment by andfoy** [comment]
@ccordoba12 @isuruf At last!

**Comment by isuruf** [comment]
Can windows dependencies be different?

File: recipes/pysyntect/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+test:
+  # Some packages might need a `test/commands` key to check CLI.
+  # List all the packages/modules that `run_test.py` imports.
+  imports:
+    - syntect
+
+about:
+  home: https://github.com/spyder-ide/pysyntect
+  license: MIT
```
Comment (isuruf) [inline recipes/pysyntect/meta.yaml:46]:
```suggestion
  license: MIT AND Apache-2.0 AND BSD-2-Clause AND BSD-3-Clause AND MPL-2.0 AND Zlib AND Unlicense
```

**Comment by andfoy** [comment]
> Can windows dependencies be different?

No, all the builds depend on GCC and LLVM

**Comment by andfoy** [comment]
> Can windows dependencies be different?

Is there any conflicting combination?

**Comment by andfoy** [comment]
@isuruf, any other comment on this one?

**Comment by isuruf** [comment]
> Can windows dependencies be different?

What I meant from this is that we only check that the licenses exist in unix. If the rust dependencies change for windows, then running this check only on unix is not enough.

**Comment by andfoy** [comment]
> What I meant from this is that we only check that the licenses exist in unix. If the rust dependencies change for windows, then running this check only on unix is not enough.

Sure! Let me check this one as well

**Comment by andfoy** [comment]
@isuruf There are no Windows-only packages

**Comment by andfoy** [comment]
@isuruf Any update on this one?

**Comment by andfoy** [comment]
@conda-forge/staged-recipes All the licenses are up-to-date, does this package is ready to merge?

**Comment by ccordoba12** [comment]
@andfoy, please squash all commits and leave a single one here.

@isuruf, we need this to be able to integrate pysyntect with Spyder and @andfoy has worked very hard to meet your review. I think this is ready.

**Comment by isuruf** [comment]
@ccordoba12, I'll need to carefully review the licenses of this PR before I can sign off on this.  (For eg: there are Apache licenses without copyright notices. If the package does have those, we need to copy them here. I currently don't have the time to do it and please note that my time here is entirely voluntary.

cc @conda-forge/staged-recipes

**Comment by ccordoba12** [comment]
> I'll need to carefully review the licenses of this PR before I can sign off on this.

Ok, I didn't know that was the main issue here. Are you referring to the licenses of all crates the Rust `syntect` library depends on?

If that's the case, we could give you a hand with that and list here all those licenses.

> For eg: there are Apache licenses without copyright notices

I think @andfoy mentioned all packages licensed as Apache are dual-licensed so that's no longer required. However, I'll double check with him if that's the case or not.

> I currently don't have the time to do it and please note that my time here is entirely voluntary.

I understand, and thanks for your time.

**Comment by isuruf** [comment]
@andfoy, can you confirm that all licenses are dual licensed MIT and Apache? I still see some Apache only licenses.

**Comment by andfoy** [comment]
> @andfoy, can you confirm that all licenses are dual licensed MIT and Apache? I still see some Apache only licenses.

There are some Apache-only licenses indeed, however, they have their respective copyright notices, as far as I know

**Comment by isuruf** [comment]
For eg: recipes/pysyntect/syntax_licenses/NSIS.txt doesn't have a copyright notice.

**Comment by andfoy** [comment]
> For eg: recipes/pysyntect/syntax_licenses/NSIS.txt doesn't have a copyright notice.

The copyright notice for that one in particular is found at the end of the full license: https://github.com/github/linguist/blob/master/vendor/licenses/grammar/NSIS.txt

**Comment by andfoy** [comment]
@isuruf, any update on this one?

**Comment by isuruf** [comment]
Thanks

**Comment by ccordoba12** [comment]
Thanks a lot @isuruf!

**Comment by andfoy** [comment]
Thanks @isuruf for your help on this one!

**Pull request merged**